docs/Toy/toy/dialect.py

```diff
@@ -0,0 +1,267 @@
+    callable SSA-region operations that contain toy computations.
+
+    Example:
+
+    ```mlir
+    toy.func @main() {
+      %0 = toy.constant dense<5.500000e+00> : tensor<f64>
+      %1 = toy.reshape(%0 : tensor<f64>) to tensor<2x2xf64>
+      toy.print %1 : tensor<2x2xf64>
+      toy.return
+    }
+    ```
+    """
+    name = 'toy.func'
+    body: Region
+    sym_name: OpAttr[StringAttr]
+    function_type: OpAttr[FunctionType]
+    sym_visibility: OptOpAttr[StringAttr]
+
+    @staticmethod
+    def from_region(name: str, ftype: FunctionType, region: Region):
+        return FuncOp.create(attributes={
+            "sym_name": StringAttr.from_str(name),
+            "function_type": ftype,
+            "sym_visibility": StringAttr.from_str("private"),
+        },
+                             regions=[region])
+
+    @staticmethod
+    def from_callable(name: str,
+                      input_types: List[Attribute],
+                      return_types: List[Attribute],
+                      func: Block.BlockCallback,
+                      private: bool = False):
+        type_attr = FunctionType.from_lists(input_types, return_types)
+        attributes = {
+            "sym_name": name,
+            "function_type": type_attr,
+        }
+        if private:
+            attributes["sym_visibility"] = "private"
+        return FuncOp.build(attributes=attributes,
+                            regions=[
+                                Region.from_block_list(
+                                    [Block.from_callable(input_types, func)])
+                            ])
+
+
+@irdl_op_definition
+class GenericCallOp(Operation):
+    name: str = "toy.generic_call"
+    arguments: Annotated[VarOperand, AnyAttr()]
+    callee: OpAttr[FlatSymbolRefAttr]
+
+    # Note: naming this results triggers an ArgumentError
+    res: Annotated[VarOpResult, AnyTensorTypeF64]
+
+    @classmethod
+    def get(cls: type[GenericCallOp], callee: Union[str, FlatSymbolRefAttr],
+            operands: List[Union[SSAValue, OpResult]],
+            return_types: List[Attribute]) -> GenericCallOp:
+        if isinstance(callee, str):
+            callee = FlatSymbolRefAttr.from_str(callee)
+
+        return super().create(operands=operands,
+                              result_types=return_types,
+                              attributes={"callee": callee})
+
+
+@irdl_op_definition
+class MulOp(Operation):
+    """
+    The "mul" operation performs element-wise multiplication between two
+    tensors. The shapes of the tensor operands are expected to match.
+    """
+    name: str = 'toy.mul'
+    arguments: Annotated[VarOperand, AnyTensorTypeF64]
+    res: Annotated[OpResult, AnyTensorTypeF64]
+
+    @classmethod
+    def from_summands(cls: type[MulOp], lhs: SSAValue, rhs: SSAValue) -> MulOp:
+        return super().create(result_types=[lhs.typ], operands=[lhs, rhs])
+
+    def verify_(self):
+        [arg0, arg1] = self.arguments
```

Comment:
This does not respect the variadic nature of the operation. IMO, this should be solved through a for loop.

Reply:
- [ ] variadic verify